compiler/src/main/scala/org/scalajs/core/compiler/GenJSCode.scala

```diff
@@ -4601,7 +4631,103 @@ abstract class GenJSCode extends plugins.PluginComponent
           patchedBody,
           allActualCaptures)
 
-      JSFunctionToScala(closure, params.size)
+      // Wrap the closure in the appropriate box for the SAM type
+      val funSym = originalFunction.tpe.typeSymbolDirect
+      if (isFunctionSymbol(funSym)) {
+        /* This is a scala.FunctionN. We use the existing AnonFunctionN
+         * wrapper.
+         */
+        JSFunctionToScala(closure, params.size)
+      } else {
+        /* This is an arbitrary SAM type (can only happen in 2.12).
+         * We have to synthesize a class like LambdaMetaFactory would do on
+         * the JVM.
+         */
+        val sam = originalFunction.attachments.get[SAMFunctionCompat].fold[Symbol] {
+          abort(s"Cannot find the SAMFunction attachment on $originalFunction at $pos")
+        } {
+          _.sam
+        }
+
+        val samWrapperClassName = synthesizeSAMWrapper(funSym, sam)
+        js.New(jstpe.ClassType(samWrapperClassName), js.Ident("init___O"),
+            List(closure))
+      }
+    }
+
+    private def synthesizeSAMWrapper(funSym: Symbol, sam: Symbol)(
+        implicit pos: Position): String = {
+      val intfName = encodeClassFullName(funSym)
+
+      val suffix = {
+        val newCount = generatedSAMWrapperCount.value + 1
+        generatedSAMWrapperCount.value = newCount
+        "$$Lambda$" + newCount // LambdaMetaFactory names classes like this
+      }
+      val generatedClassName = encodeClassFullName(currentClassSym) + suffix
```

Comment:
No, that's not the same thing. `currentClass` is the enclosing class. `intfName` is the target interface to implement.